File: leetcode/MinimumPathSum/minPathSum.java
public class Solution {
    public int minPathSum(int[][] grid) {
        // Start typing your Java solution below
        // DO NOT write main() function
        int height = grid.length;
        if (height == 0) return 0;
        int width = grid[0].length, sol[][] = new int[height][width];
        for (int i = 0; i < height; i++) {
            for (int j = 0; j < width; j++) {
                if (i == 0 && j == 0) sol[i][j] = grid[i][j];
                else 
                    sol[i][j] = grid[i][j] + Math.min(i == 0 ? Integer.MAX_VALUE : sol[i - 1][j], j == 0 ? Integer.MAX_VALUE : sol[i][j - 1]);
            }
        }
        return sol[height - 1][width - 1];
    }
}

File: leetcode/MinimumPathSum/minPathSum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m = grid.size();
        if (m == 0) return 0;
        int n = grid[0].size();
        vector<int> sol(n, 0);
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int val = !i && !j ? 0 : (!i ? sol[j - 1] : (!j ? sol[j] : min(sol[j], sol[j - 1])));
                sol[j] = val + grid[i][j];
            }
        }
        return sol[n - 1];
    }
};
